plugins/Utils/windowstatestorage.h

```diff
@@ -47,25 +84,25 @@ class WindowStateStorage: public QObject
     WindowStateStorage(const QString& dbName = nullptr, QObject *parent = nullptr);
     virtual ~WindowStateStorage();
 
-    Q_INVOKABLE void saveState(const QString &windowId, WindowState state);
+    Q_INVOKABLE void saveState(const QString &windowId, WindowStateStorage::WindowState state);
     Q_INVOKABLE WindowState getState(const QString &windowId, WindowState defaultValue) const;
 
-    Q_INVOKABLE void saveGeometry(const QString &windowId, const QRect &rect);
     Q_INVOKABLE QRect getGeometry(const QString &windowId, const QRect &defaultValue) const;
 
-    Q_INVOKABLE void saveStage(const QString &appId, int stage);
     Q_INVOKABLE int getStage(const QString &appId, int defaultValue) const;
 
     Q_INVOKABLE Mir::State toMirState(WindowState state) const;
 
-private:
-    void initdb();
+    Q_PROPERTY (const QString dbName READ getDbName)
 
-    void saveValue(const QString &queryString);
-    QSqlQuery getValue(const QString &queryString) const;
+    const QString getDbName();
 
-    static QMutex s_mutex;
+Q_SIGNALS:
+    void saveStage(const QString &appId, int stage);
+    void saveGeometry(const QString &windowId, const QRect &rect);
 
-    // NB: This is accessed from threads. Make sure to mutex it.
-    QSqlDatabase m_db;
+private:
+    QThread m_thread;
+    AsyncQuery *m_asyncQuery;
```

Comment:
Would like to see a std::shared_ptr here, else you need to make sure it gets destroyed after the class is deconstructed. but we should avoid using raw ptrs. (even if we use .get())

Reply:
See https://github.com/ubports/unity8/pull/356/files#diff-95c2b40a9d9f8f86e9324b7ded598588aebd8ef27d61c36021d3e538e492e0c9R232, which will ensure the object is deleted when the thread stops.

Reply:
Yes, here using deleteLater is safer, since the object is in another thread